storage/pebble/register.go

```diff
@@ -0,0 +1,181 @@
+package pebble
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/pebble/registers"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/onflow/atree"
+
+	"github.com/onflow/flow-go/fvm/evm/types"
+)
+
+var (
+	// MinLookupKeyLen defines the minimum length for a valid lookup key
+	//
+	// Lookup keys use the following format:
+	//     [key] / [height]
+	// Where:
+	// - key: optional variable length field
+	// - height: 8 bytes representing the block height (uint64)
+	// - separator: '/' is used to separate variable length field
+	//
+	// Therefore the minimum key would be 1 byte + # of bytes for height
+	//     / [height]
+	MinLookupKeyLen = 1 + registers.HeightSuffixLen
+)
+
+var _ types.BackendStorage = &Register{}
+
+type Register struct {
+	store  *Storage
+	height uint64
+	owner  flow.Address
+
+	batch *pebble.Batch
+}
+
+// NewRegister creates a new index instance at the provided height, all reads and
+// writes of the registers will happen at that height.
+// this is not concurrency safe.
+func NewRegister(
+	store *Storage,
+	height uint64,
+	owner flow.Address,
+	batch *pebble.Batch,
+) *Register {
+	return &Register{
+		store:  store,
+		height: height,
+		owner:  owner,
+
+		batch: batch,
+	}
+}
+
+func (r *Register) GetValue(owner, key []byte) ([]byte, error) {
+	if r.owner != flow.BytesToAddress(owner) {
+		return nil, fmt.Errorf("owner mismatch. Storage expects a single owner %s, given %s", r.owner.Hex(), flow.BytesToAddress(owner).Hex())
+	}
+
+	lookupKey := newLookupKey(r.height, key)
+	return r.lookupRegister(lookupKey.Bytes())
+}
+
+func (r *Register) SetValue(owner, key, value []byte) error {
+	if r.owner != flow.BytesToAddress(owner) {
+		return fmt.Errorf("owner mismatch. Storage expects a single owner %s, given %s", r.owner.Hex(), flow.BytesToAddress(owner).Hex())
+	}
+
+	encoded := newLookupKey(r.height, key).Bytes()
+
+	var db pebble.Writer = r.store.db
+	if r.batch != nil {
+		db = r.batch
+	}
+	err := db.Set(encoded, value, nil)
+	if err != nil {
+		return fmt.Errorf("failed to set key: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Register) ValueExists(owner, key []byte) (bool, error) {
+	val, err := r.GetValue(owner, key)
+	if err != nil {
+		return false, err
+	}
+
+	return len(val) == 0, nil
+}
+
+func (r *Register) AllocateSlabIndex(_ []byte) (atree.SlabIndex, error) {
```

Comment:
If possible, I'd like to get rid of this method from the [fvm.environment.ValueStore](https://github.com/onflow/flow-go/blob/master/fvm/environment/value_store.go#L24) interface.

I think we could probably build this method on top of `GetValue` and `SetValue`, meaning we can refactor it into a function rather than a method:

```
func AllocateSlabIndex(owner []byte, store envrionment.ValueStore) atree.SlabIndex, error) {
  cur := readSlabIndex(store, owner)
  next := cur + 1
  storeSlabIndex(store, owner, cur)
  return next
}
```